models/cost_per_serverless_task_run.sql

```diff
@@ -0,0 +1,151 @@
+
+    {% if is_incremental() %}
+    from table(generator(rowcount => (24 * 7)))
+    {% else %}
+    from table(generator(rowcount => (24 * 730)))
+    {% endif %}
+),
+
+-- 1 row per hour a run ran
+run_hours as (
+    select
+        hours_list.hour_start,
+        hours_list.hour_end,
+        runs.*
+    from hours_list
+    inner join serverless_runs as runs
+        on hours_list.hour_start >= date_trunc('hour', runs.execution_start_time)
+            and hours_list.hour_start < runs.end_time
+),
+
+run_seconds_per_hour as (
+    select
+        *,
+        datediff('millisecond', greatest(execution_start_time, hour_start), least(end_time, hour_end)) as num_milliseconds_query_ran,
+        sum(num_milliseconds_query_ran) over (partition by task_fqn_name, hour_start) as total_query_milliseconds_in_hour,
+        div0(num_milliseconds_query_ran, total_query_milliseconds_in_hour) as fraction_of_total_query_time_in_hour,
+        hour_start as hour
+    from run_hours
+),
+
+credits_billed_hourly as (
+    select
+        start_time as hour,
+        database_name,
+        database_id,
+        schema_name,
+        schema_id,
+        task_name,
+        credits_used as credits_used_serverless_task
+    from {{ ref('stg_serverless_task_history') }}
+    group by all
+),
+
+run_cost as (
+    select
+        run_seconds_per_hour.*,
+        credits_billed_hourly.credits_used_serverless_task * run_seconds_per_hour.fraction_of_total_query_time_in_hour as allocated_serverless_task_credits_in_hour,
+        allocated_serverless_task_credits_in_hour * daily_rates.effective_rate as allocated_serverless_task_cost_in_hour
+    from run_seconds_per_hour
+    inner join credits_billed_hourly
+        on run_seconds_per_hour.database_id = credits_billed_hourly.database_id
+            and run_seconds_per_hour.schema_id = credits_billed_hourly.schema_id
+            and run_seconds_per_hour.task_name = credits_billed_hourly.task_name
+            and run_seconds_per_hour.hour = credits_billed_hourly.hour
+    inner join {{ ref('daily_rates') }} as daily_rates
+        on date(run_seconds_per_hour.start_time) = daily_rates.date
+            and daily_rates.service_type = 'SERVERLESS_TASK'
+            and daily_rates.usage_type = 'serverless tasks'
+),
+
+cost_per_run as (
+    select
+        query_id,
+        any_value(start_time) as start_time,
+        any_value(end_time) as end_time,
+        any_value(task_database_id) as task_database_id,
+        any_value(task_database_name) as task_database_name,
+        any_value(task_schema_id) as task_schema_id,
+        any_value(task_schema_name) as task_schema_name,
+        any_value(task_name) as task_name,
+        any_value(execution_start_time) as execution_start_time,
+        sum(allocated_serverless_task_credits_in_hour) as serverless_task_credits,
+        sum(allocated_serverless_task_cost_in_hour) as serverless_task_cost,
+    from run_cost
+    group by 1
+)
```

Comment:
if we dedupe `credits_billed_hourly` as per my comment above, is this aggregation here necessary?